### PR TITLE
Implement diff algorithm for Hashable types

### PR DESCRIPTION
**Pull request by vadymmarkov**
This PR implements a new custom algorithm to get the difference between two arrays of `Hashable` types.

```swift
// With custom comparator (for Equatable types)
Diffuse.diff(old: old, new: new, comparator: { $0 == $1 })
// NEW: for Hashable types
Diffuse.diff(old: old, new: new)
```

The algorithm can be described in the following steps:
1. Early return if either the new or the old array is empty.
2. Map both arrays to sets of custom `<Item<T>` type.
3. Get an array with the elements that are either in the new set or in the old set, but not in both.
4. Sort by offset where old items come first.
5. Iterate over the array of differences.
5.1. **MOVE** - if the given value exists both in the new and the old sets of changes.
5.2. **REMOVE** - Assume that the old item has been removed.
6. Iterate over the array of new items (extracted from the array of differences).
6.1. **UPDATE** - if the given offset exists both in the new and the old sets of changes.
6.2. **INSERT** - if neither value nor offset exists in the old set of changes.

I know it's not super easy to follow 😄, but basically it's based on `symmetricDifference` function from `SetAlgebra` protocol. All the extra dictionaries and arrays are used to optimize performance by reducing the number of iterations.

I also did some benchmarks and this algorithm seems to be faster than other Swift diffing libraries, which is kind of surprising and suspicious at the same time. But the tests are green, so 🤞.

### CODE REVIEW

File: Diffuse/Shared/Diffuse.swift [hunk 1]
```diff
@@ -65,4 +59,74 @@ public struct Diffuse<T> where T: Equatable {
...
+        // 1. Early return if either the new or the old array is empty.
+        if new.isEmpty {
+            let changes = old.enumerated().map({ offset, _ in return Change.remove(row: offset) })
+            return CollectionChanges(allChanges: changes)
+        } else if old.isEmpty {
+            let changes = new.enumerated().map({ offset, _ in return Change.insert(row: offset) })
+            return CollectionChanges(allChanges: changes)
+        }
+
+        // 2. Map both arrays to sets of `<Item<T>` type.
```
Comment (bstien) [inline Diffuse/Shared/Diffuse.swift]:
```suggestion
        // 2. Map both arrays to sets of `Item<T>` type.
```
I guess this is a typo

**Comment by shredlocker** [comment]
Have you tested this algorithm against a multi threaded version of the one Bastian did? I think threads in this problem could be a cool

**Comment by vadymmarkov** [comment]
@shredlocker Is there a multi-threaded version?

**Comment by shredlocker** [comment]
A hypothetical version :P

**Review by bstien — APPROVED** [review]
Really nice work 😄🎉🔥

**Pull request merged**